### DISCUSSION TITLE
Fixing the telemetry mess

### DISCUSSION THREAD

**Discussion by Eugeny** [audacity/audacity#901] [Ideas & Feature Requests]
# Audacity metrics: this hot mess and how to fix it

## Preface

You've had a great idea, but the plan was flawed, poorly thought out and generally a PR clusterfuck.

Now the only right thing to do is to fix it.

The main cause of this Hindenburg is the design of the metrics system - what was going to be collected, when, how it was going to be processed and how it was going to be stored.

## Important note

OSS has a Don Quixote infestation. A lot of people have little clue about what UX is, but a _very vocal_ opinion about "telemetry" and "big corps". Usually they also fail to note that the Linux kernel they use to write these comments mostly consists of code produced by Big Software.

Right now, it's important that you don't succumb to the onslaught of white-armored knights and shelve the idea completely. Instead, it's important to distinguish between genuine criticism and the vocal minority's hysterical whining about Russian corporations using the list of filters they used to edit the sound of their farts.

## Why the plan was flawed

* It was an sensitive change that warrants a community announcement, yet went through the usual PR workflow (committed and approved internally).
* It didn't specify that telemetry was strictly opt-in.
* The data was being split between two analytics providers with no clear reason given.
* Both analytics providers are known for their poor taste in handling personal data.
* Hashing IPs is bullshit, and adding a random seed doesn't help at all because the seed itself becomes the single point of security failure.

## Why the plan wasn't as bad as some people think

* Adding a dependency on networking is not inherently a problem. Depending on libcurl is _not_ considered harmful.
* You can safely ignore people bitching about their IP being stored when checking for updates. That's was every webserver in the world does by default, it's called "access logs".
* You have to be insane to think that any sort of large corp would bother to import and use the unusable dataset that can be compiled from your IP address and the list of plugins you use. FWIW if you've ever googled anything, they already know more about you than this.

### Design

* Provide a way to people to easily and completely remove the metrics code, compile-time.
* Add opt-in prompt, and don't activate metrics (not just submission, but collection itself) until the user opts in.
* Keep the data to the minimum. You don't need their IP, you just need a unique identifier to group the events together.
* The unique identifier usage should be an opt-in as well.
* Users should have full access to the data, that means:
  * A way to preview the "regular" part of the metrics at will - i.e., the OS and app info, plugin usage stats, etc.
  * A way to preview the crash reports before submission.
* Use a fully owned environment for metrics storage and analysis.
* Proactively sanitize the data, removing any identifying information like file paths, usernames and addresses.
  
### Techicalities

* The FOSS options you have for app event analysis are few and poor, but you can start with a small webservice to log the data and Excel or R.
* Opensourcing this webservice is a sign of good will.
* Have a clear retention policy, meaning you delete stuff at a defined point in time.
* Do not use automatic report submission in Sentry or whatever crash reporter you'd be using, instead save the report to be reviewed submitted on next app start.
* Obviously verify the connection and encrypt in-flight
* Add a `#define` that completely eliminates all metrics code
* For error reporting, use the Sentry SDK but point it to your own webservice. Also, filter the data on client side first.

### Community involvement

Build a plan and post it up. You'll never get everybody to agree on anything, but what important is the outreach and open discussion. I'm sure a lot of people will come up with improvements, and there will also always be radicals for whom this is a reason enough to fork. But hey, that's what OSS is all about.

But before you do any of that, you need a public statement about the role Muse will be playing in future development of Audacity, especially in the light of their own big time dumpster fire that is the `musescore-downloader` takedown attempt.

## Expanding on this

This is a living post, I'll gladly incorporate any sensible suggestions.

**Comment by ccoenen** [comment]
Defining opinions you do not agree with as "Don Quixotes" is really not a very nice (and probably not a very correct) thing to do. And I am still opposed to tracking and telemetry. Sidestepping all the other discussions will not help this topic.

I would propose to close this. Having yet another ticket/discussion/pr/issue open will only further fragment the discussion around this topic.

**Reply by Paul-Licameli** [reply]
I have locked this discussion before it becomes too heated.

The ideas of the original post may have technical merit to be debated here, but the language was profane, and the tone is insulting to people of contrary opinions.

Please amend the wording and reopen a new discussion thread.  Let's keep the tone more professional and less confrontational.

Thank you.
